bazel: 0.22.0 -> 0.24.0

groodt opened the pull request:
###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

marsam commented:
there is already https://github.com/NixOS/nixpkgs/pull/58116

groodt commented:
> there is already #58116

Haha! I should have checked! I know @uri-canva ;)

groodt commented:
Ok, I'm going to try again.

groodt commented:
@marsam @Profpatsch
I'm looking at this line: https://github.com/NixOS/nixpkgs/pull/58147/files#diff-4746b81564a01e241a84d455575437e7R244

Would that usage of TMP cause sandbox builds to fail?

Locally, I get a different error to hydra. I get the following when running with `sandbox = true`:
`mkdir: cannot create directory '/tmp/.bazel-501': Operation not permitted`

When running with `sandbox = false` it appears to build on Darwin if I create a fake directory `/blah` with 777 chmod and change that line to use `/blah`

Hydra timing out after 600 seconds of silence, might be because the compilation takes longer than 10 minutes? I suspect that maybe hydra doesn't have the same TERM or something to see the fancy colored Bazel compilation progress. It takes at least 20-30 minutes to build.

groodt commented:
I've hacked the derivation a bit and it's almost working now. Failing at the `installCheckPhase`.

```
In file included from examples/cpp/hello-world.cc:1:
./examples/cpp/hello-lib.h:4:10: fatal error: 'string' file not found
#include <string>
         ^~~~~~~~
1 error generated.
INFO: Elapsed time: 34.407s, Critical Path: 0.16s
INFO: 0 processes.
FAILED: Build did NOT complete successfully
//examples/java-native/src/test/java/com/example/myproject:hello      NO STATUS
```

I'll continue looking at this tomorrow. In the meantime, if anybody can review my changes and suggest better ways to do it, I'm happy to modify things within reason.

groodt commented:
Note, my error above is with `sandbox = false` (I suspect it's a CPP or C toolchain issue).

If I build with `sandbox = true` I get `failed with exit code 134` so I suspect I'm doing something that is banned in sandbox mode.

uri-canva commented:
It might be because you move `.bazelrc` into `bazel_src` and then run the tests from outside it, so the tests don't pick up `.bazelrc`, which is necessary for the bazel build to pick up the right flags from nix.

groodt commented:
> It might be because you move `.bazelrc` into `bazel_src` and then run the tests from outside it, so the tests don't pick up `.bazelrc`, which is necessary for the bazel build to pick up the right flags from nix.

Good point. I'll check this. I thought that I was copying the `.bazelrc` along for the ride and then `cd ./bazel_src` would make it available. I'll try to debug.

groodt commented:
Ok, this is now building successfully on my Mac with `sandbox = false`.

With `sandbox = true` I still get `failed with exit code 134`

uri-canva commented:
Exit codes higher than 128 are usually from signals, 134-128 = 6 is `SIGABRT`. Do you know what command is aborting? Is it nix itself or one of the bazel commands?

groodt commented:
> Exit codes higher than 128 are usually from signals, 134-128 = 6 is `SIGABRT`. Do you know what command is aborting? Is it nix itself or one of the bazel commands?

Yes, I suspect the same thing.

```
Building Bazel from scratch
builder for '/nix/store/14gpkg2y5fg7my3kc5fpjm0qfcbm22vz-bazel-0.23.2.drv' failed with exit code 134
error: build of '/nix/store/14gpkg2y5fg7my3kc5fpjm0qfcbm22vz-bazel-0.23.2.drv' failed```

groodt commented:
Patched my build to output tracing. I think this is the reason why I fail to build with `sandbox = true`.

```
...
++ test -z /nix/store/9qyzw01bnla43457iwp6l9r6fbp7gk4y-zulu1.8.0_121-8.20.0.5
++ JAVAC=/nix/store/9qyzw01bnla43457iwp6l9r6fbp7gk4y-zulu1.8.0_121-8.20.0.5/bin/javac
++ [[ -x /nix/store/9qyzw01bnla43457iwp6l9r6fbp7gk4y-zulu1.8.0_121-8.20.0.5/bin/javac ]]
+++ /nix/store/9qyzw01bnla43457iwp6l9r6fbp7gk4y-zulu1.8.0_121-8.20.0.5/bin/javac -version
++ JAVAC_VERSION='dyld: Library not loaded: /System/Library/Frameworks/Cocoa.framework/Versions/A/Cocoa
  Referenced from: /nix/store/9qyzw01bnla43457iwp6l9r6fbp7gk4y-zulu1.8.0_121-8.20.0.5/bin/javac
  Reason: no suitable image found.  Did find:
        /System/Library/Frameworks/Cocoa.framework/Versions/A/Cocoa: file system sandbox blocked stat()
        /System/Library/Frameworks/Cocoa.framework/Versions/A/Cocoa: file system sandbox blocked stat()'
+ run_atexit_handlers 134
+ local exit_code=134
...
```

Profpatsch commented:
@GrahamcOfBorg build bazel

Profpatsch commented:
Darwin build is timing out.

uri-canva commented:
There's headless versions of jdk11 and jre for Linux. If we could have headless versions of jdk8 for macOS, then the build wouldn't attempt to access Cocoa and it might work. Did something change recently with how sandboxing is enforced? I think system libraries on macOS were still being access in non-pure fashion until a couple of months ago.

groodt commented:
@Profpatsch 
> Darwin build is timing out.

I see that. What would you suggest now? This is the same as others have seen trying to get a new version of Bazel over the line. How can we get more build output? On my own Mac, I see a lot more output.

Given the local sandbox failure loading Cocoa, I think it's possible that it never built successfully on Darwin with `sandbox = true` or the sandboxing mechanisms have changed recently.

I'll keep digging, but it's looking like this might have to be changed to a linux only pkg if sandboxed Darwin builds are a blocker.

groodt commented:
Quick post to summarise the difference between failing sandbox builds on a Mac vs on Hydra.

Mac (Tracing disabled. The underlying failure is due to a javac call reaching outside the sandbox for Cocoa libraries):
```
fixing libtool script ./third_party/protobuf/3.6.1/ltmain.sh
fixing libtool script ./third_party/protobuf/3.6.1/third_party/googletest/googletest/build-aux/ltmain.sh
fixing libtool script ./third_party/protobuf/3.6.1/third_party/googletest/googlemock/build-aux/ltmain.sh
no configure script, doing nothing
preBuildPhase
building
Building Bazel from scratch
builder for '/nix/store/aqg0l592dmr8ny5xgiiv3k3vbgdxg0lz-bazel-0.23.2.drv' failed with exit code 134
error: build of '/nix/store/aqg0l592dmr8ny5xgiiv3k3vbgdxg0lz-bazel-0.23.2.drv' failed
```

Hydra (https://logs.nix.ci/?key=nixos/nixpkgs.58147&attempt_id=2a30e2bb-eabb-41f2-8786-a53b3be2509a):
```
fixing libtool script ./third_party/protobuf/3.6.1/ltmain.sh
fixing libtool script ./third_party/protobuf/3.6.1/third_party/googletest/googlemock/build-aux/ltmain.sh
fixing libtool script ./third_party/protobuf/3.6.1/third_party/googletest/googletest/build-aux/ltmain.sh
no configure script, doing nothing
preBuildPhase
building
building of '/nix/store/14gpkg2y5fg7my3kc5fpjm0qfcbm22vz-bazel-0.23.2.drv' timed out after 600 seconds of silence
error: build of '/nix/store/14gpkg2y5fg7my3kc5fpjm0qfcbm22vz-bazel-0.23.2.drv' failed
```

Note the missing "Building Bazel from scratch" output in Hydra. That line comes from `./bazel_src/compile.sh`. Either we are losing logging somehow or buildPhase isn't even starting.

I'm going to try enable tracing and kick off a build, but I'm unsure if the bot will listen to me.

groodt commented:
@GrahamcOfBorg build bazel

groodt commented:
Just in case you're curious, even if I patch away the line in the build where it initially sets JAVAC_VERSION to hardcode "javac 1.8" it still fails with `sandbox = true`. This is failing with the core invocation of javac to compile the Bazel java classes.

```
/nix/store/9qyzw01bnla43457iwp6l9r6fbp7gk4y-zulu1.8.0_121-8.20.0.5/bin/javac -classpath third_party/truth8/truth-java8-extension-0.42.jar:third_party/javax_annotations/javax.annotation-api-1.3.2.jar:third_party/javax_annotations/javax.annotation-api-1.3.2-sources.jar
<SNIP ... />
-sourcepath src/java_tools/singlejar/java/com/google/devtools/build/zip:src/main/java:src/tools/xcode-common/java/com/google/devtools/build/xcode/common:src/tools/xcode-common/java/com/google/devtools/build/xcode/util:tools/java/runfiles:third_party/java/dd_plist/java:/private/var/folders/0f/v0lxnbcx4v94rbhtz37d6f2m0000gn/T/nix-build-bazel-0.23.2.drv-0/bazel_S1B9kq7r/src -d /private/var/folders/0f/v0lxnbcx4v94rbhtz37d6f2m0000gn/T/nix-build-bazel-0.23.2.drv-0/bazel_S1B9kq7r/classes -source 1.8 -target 1.8 -encoding UTF-8 @/private/var/folders/0f/v0lxnbcx4v94rbhtz37d6f2m0000gn/T/nix-build-bazel-0.23.2.drv-0/bazel_nuJxfgmI/param
dyld: Library not loaded: /System/Library/Frameworks/Cocoa.framework/Versions/A/Cocoa
  Referenced from: /nix/store/9qyzw01bnla43457iwp6l9r6fbp7gk4y-zulu1.8.0_121-8.20.0.5/bin/javac
  Reason: no suitable image found.  Did find:
        /System/Library/Frameworks/Cocoa.framework/Versions/A/Cocoa: file system sandbox blocked stat()
        /System/Library/Frameworks/Cocoa.framework/Versions/A/Cocoa: file system sandbox blocked stat()
++ exit 0
+ run_atexit_handlers 0
+ local exit_code=0
+ local failed=no
+ for handler in ${ATEXIT_HANDLERS}
```

Apologies for all the spamming.

The above indicates to me that it is not possible to compile Java applications on Darwin with `sandbox = true` unless there is a way to prevent javac from reaching out for Cocoa. I hope that I'm mistaken. Does anybody know of any other Java applications that are building on Darwin with sandboxing enabled?

If not, I think this leaves Bazel 0.23.2 on Darwin dead in the water as an official Nixpkg unless a headless jdk comes along for Darwin.

uri-canva commented:
0.24.0 was just released. @Profpatsch any ideas how we can debug this issue? It looks like the issues we get when building locally on macOS are completely different from building in hydra on macOS to be honest, I'm not sure spending time fixing what we're encountering locally is even helpful.

groodt commented:
I can close this one if people wish to try 0.24.0. I've just attempted it and it's failing to compile for me on OSX, so I don't think it's ready just yet.

> I'm not sure spending time fixing what we're encountering locally is even helpful.
I think it would be great to see why Hydra isn't outputting any logs once it starts the build. My current branch has tracing enabled, so if we can get that to run on Hydra, we'll at least see where it stops.

I agree that the Hydra Darwin builds outputting no logs could be a different issue / PR though. That assumes it's a general issue with Hydra and Darwin and not something specific to building Bazel.

groodt commented:
I've quickly hacked it to build 0.24.0. Same issue as 0.23.2. It builds fine for me with sandbox=false. Fails to build with sandbox=true.

groodt commented:
@Profpatsch I'd appreciate it if you could kick off another build so that we can see the trace output (if any) on Hydra.

groodt commented:
Ready for review and Hydra builds. I've left TODOs with the tracing code which can be removed after testing Darwin builds on Hydra.

groodt commented:
@Profpatsch Sorry to bother, any chance you could kick off a build for this please? Or tag somebody else with appropriate permissions.

I'd like to confirm why Hydra builds are failing.

Profpatsch commented:
Oh sorry, missed this notification. Sure

@GrahamcOfBorg build bazel

groodt commented:
Update on the situation. It appears that both Hydra and my local `sandbox=true` builds are failing at the same place. The error just manifests itself differently.

Tracing output from Hydra:
```
++ test -z /nix/store/9qyzw01bnla43457iwp6l9r6fbp7gk4y-zulu1.8.0_121-8.20.0.5
++ JAVAC=/nix/store/9qyzw01bnla43457iwp6l9r6fbp7gk4y-zulu1.8.0_121-8.20.0.5/bin/javac
++ [[ -x /nix/store/9qyzw01bnla43457iwp6l9r6fbp7gk4y-zulu1.8.0_121-8.20.0.5/bin/javac ]]
+++ /nix/store/9qyzw01bnla43457iwp6l9r6fbp7gk4y-zulu1.8.0_121-8.20.0.5/bin/javac -version
building of '/nix/store/qqiyp494jprdxbbak4z6w5zg68kbw3vm-bazel-0.24.0.drv' timed out after 600 seconds of silence
error: build of '/nix/store/qqiyp494jprdxbbak4z6w5zg68kbw3vm-bazel-0.24.0.drv' failed
```

Compare that to my failure on Mac:
```
++ test -z /nix/store/9qyzw01bnla43457iwp6l9r6fbp7gk4y-zulu1.8.0_121-8.20.0.5
++ JAVAC=/nix/store/9qyzw01bnla43457iwp6l9r6fbp7gk4y-zulu1.8.0_121-8.20.0.5/bin/javac
++ [[ -x /nix/store/9qyzw01bnla43457iwp6l9r6fbp7gk4y-zulu1.8.0_121-8.20.0.5/bin/javac ]]
+++ /nix/store/9qyzw01bnla43457iwp6l9r6fbp7gk4y-zulu1.8.0_121-8.20.0.5/bin/javac -version
++ JAVAC_VERSION='dyld: Library not loaded: /System/Library/Frameworks/Cocoa.framework/Versions/A/Cocoa
  Referenced from: /nix/store/9qyzw01bnla43457iwp6l9r6fbp7gk4y-zulu1.8.0_121-8.20.0.5/bin/javac
  Reason: no suitable image found.  Did find:
        /System/Library/Frameworks/Cocoa.framework/Versions/A/Cocoa: file system sandbox blocked stat()
        /System/Library/Frameworks/Cocoa.framework/Versions/A/Cocoa: file system sandbox blocked stat()'
+ run_atexit_handlers 134
+ local exit_code=134
```

You can see that both fail when running `javac -version` inside the sandbox. Locally it fails with SIGABRT and on Hydra it just hangs for some reason. I'm guessing that the way that the Hydra sandbox environment is setup for Darwin, means that loading native libraries like this might be blocked at a low-level or the same SIGABRT is happening, but not being propagated to the logs somehow.

Any suggestions @Profpatsch ? We now know what the error is and can reproduce it locally and on Hydra. This is good progress I think.

Any ideas how this ever could have worked in a sandbox?

uri-canva commented:
`javac` is linked to Cocoa so I guess we need to use `install_name_tool` or build the jdk from source:

```
MacBook-Pro-6:nixpkgs uri$ otool -L /nix/store/9qyzw01bnla43457iwp6l9r6fbp7gk4y-zulu1.8.0_121-8.20.0.5/bin/javac
/nix/store/9qyzw01bnla43457iwp6l9r6fbp7gk4y-zulu1.8.0_121-8.20.0.5/bin/javac:
	/System/Library/Frameworks/Cocoa.framework/Versions/A/Cocoa (compatibility version 1.0.0, current version 19.0.0)
	/System/Library/Frameworks/Security.framework/Versions/A/Security (compatibility version 1.0.0, current version 55179.11.0)
	/System/Library/Frameworks/ApplicationServices.framework/Versions/A/ApplicationServices (compatibility version 1.0.0, current version 45.0.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.5)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 169.3.0)
	/usr/lib/libobjc.A.dylib (compatibility version 1.0.0, current version 228.0.0)
	/System/Library/Frameworks/Foundation.framework/Versions/C/Foundation (compatibility version 300.0.0, current version 945.16.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 744.18.0)
```

groodt commented:
> `javac` is linked to Cocoa so I guess we need to use `install_name_tool` or build the jdk from source:

Yes. That seems like it should be done as part of the jdk nixpkg. I guess there is a reason that it's not been done.

I guess another alternative would be to use the Bazel binary distributions of Bazel (at least for Darwin). I see that Ant, Gradle, Maven and SBT are all using binary distributions. As a result, they don't execute `javac` inside the sandbox.

I guess building from source is the Nix preference, but there are official binary Linux, Darwin and Windows builds of Bazel available.

uri-canva commented:
I had completely forgotten about `extra-sandbox-paths`: https://github.com/NixOS/nix/issues/2732#issuecomment-474150868

It's possible the value of that setting on hydra changed recently, but on master `bazel` is ok because it was built from before the change, so it's still cached.

@LnL7 Do you know where I can look up the value of that setting on hydra.nixos.org? I couldn't find anything in https://github.com/NixOS/nixos-org-configurations.

groodt commented:
> on master `bazel` is ok because it was built from before the change, so it's still cached.

I'm trying to confirm if 0.22 still builds on master now with: https://github.com/NixOS/nixpkgs/pull/58557

Just needs someone to kick-off a Hydra build.

Good tip about `extra-sandbox-paths`

groodt commented:
Looks like `extra-sandbox-paths` almost does the trick. Bazel 0.24.0 gets right to the end of the build with:

```
sandbox = true
extra-sandbox-paths = /System/Library/Frameworks /System/Library/PrivateFrameworks /usr/lib
```

What is the policy for setting those values on Hydra? Or are those settings not permitted on Hydra?

It now fails with:
```
Starting local Bazel server and connecting to it...
Server crashed during startup. Now printing /private/var/folders/0f/v0lxnbcx4v94rbhtz37d6f2m0000gn/T/nix-build-bazel-0.24.0.drv-0/_bazel_groodt/561765db2a1d1b8606e4b7b3fd145d4b/server/jvm.out
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.protobuf.UnsafeUtil (file:/private/var/folders/0f/v0lxnbcx4v94rbhtz37d6f2m0000gn/T/nix-build-bazel-0.24.0.drv-0/_bazel_groodt/install/74381fd39e30d52ac531dc05c55dda59/_embedded_binaries/A-server.jar) to field java.nio.Buffer.address
WARNING: Please consider reporting this to the maintainers of com.google.protobuf.UnsafeUtil
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
I/O Error: Failed to bind
```

That's Bazel trying to start a local server and failing to connect to a port I think. Another sandboxing change?

LnL7 commented:
@uri-canva None of the darwin machines used by hydra currently have sandboxing enabled.

groodt commented:
@LnL7 Would you mind kicking off this build on Darwin please: https://github.com/NixOS/nixpkgs/pull/58557

etu commented:
@GrahamcOfBorg build bazel

LnL7 commented:
Going through the comments I see a bunch of mostly unrelated comments and I see a bunch of unrelated issues that's probably confusion.

Firstly sandboxing is currently not fully functional on darwin and not enabled by default, I have been spending some time to get it back on track but it's not a requirement for merge requests. Hydra and ofborg are 2 different things. Hydra only builds reviewed changes and these will be pushed to the cache eg. https://hydra.nixos.org/job/nixpkgs/trunk/bazel.x86_64-darwin, while ofborg runs for merge requests and uses different hardware. I enabled partial darwin sandboxing there to get an overview of what still needs fixing, like this case.

There are also some functional differences between sandboxing on linux vs darwin.
- `/tmp` is global and not namespaced per build, which means if 2 consecutive builds try to write the same file there without cleaning it up nicely the second will fail because builds run as different unprivileged users. It's better to make the build use `NIX_BUILD_TOP` instead since that's managed by nix and will be cleaned up automatically regardless if the build fails or not.
- Similarly localhost is not namespaced either, if multiple builds try to listen on the same port at the same time one of them will fail. Because of this the nix sandbox needs explicit whitelisting of builds that are allowed to do this with `__darwinAllowLocalNetworking = true;`.

etu commented:
@GrahamcOfBorg build bazel

groodt commented:
> ofborg runs for merge requests and uses different hardware. I enabled partial darwin sandboxing there to get an overview of what still needs fixing, like this case.

This explains it! Thank you! I've been pulling my hair out trying to figure out how 0.22 could ever have succeeded with sandboxing enabled on Darwin. The answer is that it wouldn't. It would fail just as it fails on my local machine.

@LnL7 do you have any insights into why these Darwin failures with ofberg manifest as "time outs" rather than actual sandbox errors which they are likely to be?

uri-canva commented:
> I enabled partial darwin sandboxing there to get an overview of what still needs fixing, like this case.

Is the configuration available anywhere? Nothing jumps to my eye in https://github.com/NixOS/ofborg.

We're trying to repro the failure in ofborg on our laptops but we can't get the same errors. At least we now know it's a sandbox error due to jdk linking in all system libraries and we also know how come it doesn't cause any issues in hydra, but since neither of us can trigger darwin builds on ofborg it's still very hard for us to debug the derivation.

uri-canva commented:
If I have sandbox enabled, and change the reference to the link section with `install_name_tool`, I get:

```
+++ /nix/store/fn6gwllqy98qhmizikmkzdfmpq05nyqr-zulu1.8.0_121-8.20.0.5/bin/javac -version
++ JAVAC_VERSION='dyld: Library not loaded: /nix/store/hdsb3b20czzy8424yzhx3xkfyr11mmh7-apple-framework-Cocoa/Library/Frameworks/Cocoa.framework/Cocoa
  Referenced from: /nix/store/fn6gwllqy98qhmizikmkzdfmpq05nyqr-zulu1.8.0_121-8.20.0.5/bin/javac
  Reason: no suitable image found.  Did find:
        /nix/store/hdsb3b20czzy8424yzhx3xkfyr11mmh7-apple-framework-Cocoa/Library/Frameworks/Cocoa.framework/Cocoa: file system sandbox blocked stat()
        /System/Library/Frameworks/Cocoa.framework/Cocoa: file system sandbox blocked stat()'
```

The reference to the system framework might come from some other dynamic library loaded in `javac`, but it looks like even the library from nixpkgs cannot be accessed in the sandbox.

uri-canva commented:
Darwin success: https://github.com/NixOS/nixpkgs/pull/58147/checks?check_run_id=88814348

Was the sandbox disabled?

uri-canva commented on pkgs/development/tools/build-managers/bazel/default.nix:
```diff
@@ -236,34 +254,44 @@ stdenv.mkDerivation rec {
-        --prepend=scripts/bazel-complete-header.bash \
-        --prepend=scripts/bazel-complete-template.bash
+    # TODO Remove. Only in place for debugging Darwin Hydra issue.
+    set -o xtrace
```
Note this should be removed before merging.

groodt commented:
> Darwin success: https://github.com/NixOS/nixpkgs/pull/58147/checks?check_run_id=88814348
> 
> Was the sandbox disabled?

@uri-canva I imagine it must have been disabled by @LnL7 or somebody else.

In summary @uri-canva all of our struggling with Darwin specific sandbox issues was unnecessary. Still valuable information, but ultimately unnecessary. It should have been possible to merge 0.23.0 or 0.23.2 once we had them building locally on Mac and remotely with Ofborg.

Main important things I've learned:
1. Hydra and Ofborg are different running on different hardware for different purposes.
2. On Hydra, `sandbox` is not enabled for Darwin builds. Precisely because of issues like we've encountered. `sandbox` simply isn't ready yet for Darwin.
3. Ofborg builds can have `sandbox` enabled or disabled for Darwin, so it can be flaky. I think it's not intended for Ofborg to block merges to master, but rather to give higher confidence and catch obvious issues. 
4. There is an issue with Ofborg build failure messages on Darwin. When it encounters a `sandbox` related issue like we have, it is manifesting as a timeout incorrectly.

@Profpatsch I've squashed all my commits into 1 now and removed all TODOs. So the PR is ready for review.

Ironically, it is now the linux build failing. It builds Bazel and gets all the way to the tests and then fails with:
```
ERROR: /build/_bazel_nixbld/96c321f6d4067432c019ff2725c18acb/external/bazel_tools/tools/jdk/BUILD:218:1: no such package '@remote_java_tools//': java.io.IOException: Error downloading [https://mirror.bazel.build/bazel_java_tools/java_tools_pkg-0.5.1.tar.gz] to /build/_bazel_nixbld/96c321f6d4067432c019ff2725c18acb/external/remote_java_tools/java_tools_pkg-0.5.1.tar.gz: Unknown host: mirror.bazel.build and referenced by '@bazel_tools//tools/jdk:JacocoCoverageRunner'
ERROR: Analysis of target '//examples/cpp:hello-success_test' failed; build aborted: no such package '@remote_java_tools//': java.io.IOException: Error downloading [https://mirror.bazel.build/bazel_java_tools/java_tools_pkg-0.5.1.tar.gz] to /build/_bazel_nixbld/96c321f6d4067432c019ff2725c18acb/external/remote_java_tools/java_tools_pkg-0.5.1.tar.gz: Unknown host: mirror.bazel.build
```

mboes commented:
@groodt the linux build succeeds when the sandbox is turned _off_ (it already is on Darwin). The reason is that in order to be sandbox friendly, you need to prefetch dependencies, since there is no network access inside the sandbox. We already need to do this, so you probably just need to add new  tarballs to `srcDeps` (grep Bazel's `default.nix`).

groodt commented:
Status update.

Building successfully on Darwin with these settings in `/etc/nix/nix.conf`:
```
sandbox = true
extra-sandbox-paths = /System/Library/Frameworks /System/Library/PrivateFrameworks /usr/lib
```

Failing on linux. Investigating now, but it looks like a standard issue.

groodt commented:
> @groodt the linux build succeeds when the sandbox is turned _off_ (it already is on Darwin). The reason is that in order to be sandbox friendly, you need to prefetch dependencies, since there is no network access inside the sandbox. We already need to do this, so you probably just need to add new tarballs to `srcDeps` (grep Bazel's `default.nix`).

@mboes This is true somewhat, but if you look at the game of whack-a-mole I've been playing in this thread, it's not so straightforward. Ofborg has been running with sandboxing enabled, so Darwin builds have been "timing out" (actually SIGABRT sandbox errors).

This resulted in us chasing our tails and making changes to the already working derivation of 0.22. It might be that some of these changes I've made to make Darwin sandbox happy, are causing it to fail on linux.

The early builds of this derivation were passing on linux and "timed out" on Darwin. Now it is working on Darwin and failing on Linux. 👍 

Anyway, it's probably just a matter of downloading additional files or rolling back some of my changes that made Darwin sandbox happy. I'll take a look today.

groodt committed:
bazel: 0.22.0 -> 0.24.0

uri-canva commented:
One thing I thought of in the meantime: skipping `0.23.x` could make it harder for users to upgrade, since Bazel keeps only one version of backwards compatibility. We should have two separate upgrade commits, even if they're one right after the other, so that people who pin nixpkgs can upgrade to `0.23.2` first and then to `0.24.0`.

uri-canva commented:
I reopened https://github.com/NixOS/nixpkgs/pull/58116 in case we want to go ahead and merge that first.

groodt commented:
Having a Hydra build of `bazel 0.23.2` for pinning sounds good to me.

I'm currently testing linux sandboxed builds of this PR, but I think it's also ready now.

groodt commented:
This is now building and ready for review @Profpatsch @mboes Got there in the end!

Please see @uri-canva comment about thoughts of doing `0.22 -> 0.23.2` first. I don't mind if that is done first. I don't see too much benefit in sticking to an old Bazel for very long, so we should get this one merged shortly after if possible.

See below table for how I've tested this derivation on local Darwin and Linux. I'm confident that it will work on Hydra now. Linux should work on Ofborg and Hydra. Darwin might fail on Ofborg, but should succeed on Hydra given the testing that I've done.

| Status            | OS                  | sandbox       | Notes                                                                         |
| ------------- | ------------- |-------------|------------------------------------------------- |
| ✅                  | Linux              | false             | single-user Nix 2.2.1 on privileged debian:jessie |
| ✅                  | Linux              | true              | single-user Nix 2.2.1 on privileged debian:jessie |
| ✅                  | Darwin           | false             | sandbox = false                                                     |
| ✅                  | Darwin           | true              | sandbox = true<br/>extra-sandbox-paths = /System/Library/Frameworks /System/Library/PrivateFrameworks /usr/lib|

Infinisil commented:
@GrahamcOfBorg build bazel

LnL7 commented on pkgs/development/tools/build-managers/bazel/default.nix:
```diff
@@ -85,9 +90,13 @@ stdenv.mkDerivation rec {
 
+  # Necessary for the tests to pass on Darwin with sandbox enabled.
+  # Bazel starts a local server and needs to bind a local address.
+  __darwinAllowLocalNetworking = true;
```
I assume this is inherent with bazel, is this a random or fixed port? If it's static hydra and local builds will have problems if multiple bazel builds are running at the same time.

groodt replied:
Yes, it is inherent to Bazel. It uses a daemon for faster builds. There used to be a `--batch` option, but it is being removed: https://docs.bazel.build/versions/master/command-line-reference.html#flag--batch

The code that I looked at seems to allow the kernel to choose a random port.

LnL7 reviewed (approved):
Looks good from the darwin side. Thanks for fixing the sandbox issues and sorry for the confusion.

mboes commented:
@profpatsch As far as I'm concerned we can merge.

groodt commented:
@profpatsch Any feedback? There is typically a new Bazel release every 2-4
weeks, so maybe we should land this one after all. If somebody needs an
older version, they can call a custom derivation. Anybody using Bazel has
to stay up to date with new versions anyway.

On Sun, 31 Mar 2019 at 23:26, Mathieu Boespflug <notifications@github.com>
wrote:

> @profpatsch As far as I'm concerned we can merge.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/58147#issuecomment-478337330>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAU9d0uoR4vSn6NiVSrYri-lz5cAcdcYks5vcKmPgaJpZM4cEZQ6>
> .
>

Profpatsch commented:
Yep, we should merge even if we skip 0.23, the releases are too fast for the manpower we have here.

flokli commented:
Should this be backported to 19.03 as well?

groodt commented:
> Should this be backported to 19.03 as well?

I'd like this too, but it's a bit late in the game for 19.03 I think. They've been working really hard to get it out the door. I don't know what the policy is for late inclusions. I guess with Nix it should be safe, but 🤷‍♂️

Profpatsch commented:
Nah, it’s a major, incompatible update, so we shouldn’t backport it.
